gwcs/wcs.py

```diff
@@ -634,3 +641,235 @@ def fix_inputs(self, fixed):
+        #               RotateNative2Celestial(crval1, crval2, 180))
+        return hdr
+
+
+def fit_2D_poly(degree, max_error, max_rms, u, v, x, y, verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied inputs (u, v) and
+    outputs (x, y)
+    """
+    llsqfitter = LinearLSQFitter()
+    # The case of one pass with the specified polynomial degree
+    if degree is not None:
+        poly_x = Polynomial2D(degree=degree)
+        poly_y = Polynomial2D(degree=degree)
+        fit_poly_x = llsqfitter(poly_x, u, v, x)
+        fit_poly_y = llsqfitter(poly_y, u, v, y)
+        max_resid, rms = compute_distance_residual(x, y,
+                                                   fit_poly_x(u, v), fit_poly_y(u, v))
+    else:
+        cond = True
+        degree = 2
+        while cond:
+            poly_x = Polynomial2D(degree=degree)
+            poly_y = Polynomial2D(degree=degree)
+            fit_poly_x = llsqfitter(poly_x, u, v, x)
+            fit_poly_y = llsqfitter(poly_y, u, v, y)
+            max_resid, rms = compute_distance_residual(x, y,
+                                                       fit_poly_x(u, v), fit_poly_y(u, v))
+            if verbose:
+                print(f'Degree = {degree}, max_resid = {max_resid}, rms = {rms}')
+            if ((max_error is not None and max_resid < max_error and max_rms is None)
+                 or (max_error is not None and max_resid < max_error and max_rms is not None
+                     and rms < max_rms)
+                 or (max_error is None and max_rms is not None and rms > max_rms)):
+                cond = False
+                if verbose:
+                    print('terminating condition met')
+                break
+            else:
+                degree += 1
+                if degree == 10:
+                    raise RuntimeError("Fit conditions not met")
+    return fit_poly_x, fit_poly_y, max_resid, rms
+
+
+def compute_distance_residual(undist_x, undist_y, fit_poly_x, fit_poly_y):
+    """
+    Compute the distance residuals and return the rms and maximum values.
+    """
+    dist = np.sqrt((undist_x - fit_poly_x)**2 + (undist_y - fit_poly_y)**2)
+    max_resid = dist.max()
+    rms = np.sqrt((dist**2).sum() / dist.size)
+    return max_resid, rms
+
+
+def reform_poly_coefficients(fit_poly_x, fit_poly_y):
+    """
+    The fit polynomials must be recombined to align with the SIP decomposition
+
+    The result is the f(u,v) and g(u,v) polynomials, and the CD matrix.
+    """
+    # Extract values for CD matrix and recombing
+    c11 = fit_poly_x.c1_0.value
+    c12 = fit_poly_x.c0_1.value
+    c21 = fit_poly_y.c1_0.value
+    c22 = fit_poly_y.c0_1.value
+    sip_poly_x = fit_poly_x.copy()
+    sip_poly_y = fit_poly_y.copy()
+    # Force low order coefficients to be 0 as defined in SIP
+    sip_poly_x.c0_0 = 0
+    sip_poly_y.c0_0 = 0
+    sip_poly_x.c1_0 = 0
+    sip_poly_x.c0_1 = 0
+    sip_poly_y.c1_0 = 0
+    sip_poly_y.c0_1 = 0
```

Comment:
This code is unnecessary due to `if (i + j > 1) and (i + j < degree + 1)` in the double-`for` loop below.

Reply:
I take this back. This is not _just yet_ simply adding header keywords.

Reply:
Ripped out

Reply:
Oh, I can't understand what I wanted to say in my last comment above. My intention was that you do not remove this if it is being passed to another function. Sorry about that. ONLY IF you consider it is OK to remove it.